lib/gc/Transforms/GPU/LinalgToXeGPU.cpp

```diff
@@ -686,53 +720,214 @@ createDescriptorTiles(PatternRewriter &rewriter, Location loc, Value src,
+            .getResult();
+    blockOffset = rewriter.create<arith::AddIOp>(loc, rowBlockOffset, yIntOffs)
+                      .getResult();
+
+    memrefStrides = {rowStride, 1};
+    src = subView.getOperand(0);
+  } else {
+    // If the source is not a subview, then the blockOffset is 0
+    blockOffset = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+    memrefStrides = {srcType.getShape()[1], 1};
+  }
+
+  // Scatter descriptors only work with 1D memrefs
+  src = utils::flattenMemref(rewriter, loc, src);
+
+  return createScatterDescriptorTiles(
+      rewriter, loc, /*flatMemref=*/src, /*loadShape2D=*/loadShape,
+      /*tileSize2D=*/descTile, /*memrefStrides=*/memrefStrides,
+      /*blockOffset=*/blockOffset);
+}
+
+static SmallVector<Value> createDescriptorTiles(
+    PatternRewriter &rewriter, Location loc, Value src,
+    ArrayRef<int64_t> loadShape, ArrayRef<int64_t> descTile,
+    std::optional<ArrayRef<int64_t>> loadOffsets = std::nullopt,
+    int arrayLength = 1, bool transpose = false) {
+
+  if (utils::hasSharedMemSpace(src)) {
+    assert(!transpose && "Transpose is not supported for shared memory");
+    assert(arrayLength == 1 &&
+           "Array descriptors are not supported for shared memory");
+    assert(!loadOffsets && "Load offsets are not supported for shared memory");
+    return createSLMDescTiles(rewriter, loc, src, loadShape, descTile);
+  }
+  return createNdDescriptorTiles(
+      rewriter, loc, src, loadShape,
+      loadOffsets.value_or(SmallVector<int64_t>{0, 0}), descTile, arrayLength,
+      transpose);
+}
 
+SmallVector<int64_t> determine2DTileSize(ArrayRef<int64_t> totalShape,
+                                         bool isVnni, int64_t elemByteWidth) {
   // TODO: Fetch actual list of supported load configs.
   int64_t maxHeight = 32;
   int64_t maxWidth = 64 / elemByteWidth;
   // Assumes VNNI-factor 2.
   // TODO: Make the VNNI-factor flexible.
   if (isVnni)
     maxWidth /= 2;
-  int64_t maxArrayLength = 4;
-  int64_t sgLoadRows = std::min(sgTile2D[0], maxHeight);
-  int64_t sgLoadCols = std::min(sgTile2D[1], maxWidth);
-  int64_t arrayLength = std::min(maxWidth / sgLoadCols, maxArrayLength);
-  // In case of partial fit, load only single tile.
-  // NOLINTBEGIN
-  if (maxWidth % sgLoadCols != 0 || arrayLength != 4 || arrayLength != 2)
-    arrayLength = 1;
 
-  // TODO: Add variable array_length support.
-  arrayLength = 1;
-  // NOLINTEND
+  int64_t sgLoadRows = std::min(totalShape[0], maxHeight);
+  int64_t sgLoadCols = std::min(totalShape[1], maxWidth);
 
-  return createDescriptorTiles(rewriter, loc, src, sgTile2D, {0, 0},
-                               {sgLoadRows, sgLoadCols}, arrayLength,
-                               transpose);
+  return SmallVector<int64_t>{sgLoadRows, sgLoadCols};
+}
+
+// Create coarse sub-tiles to be loaded by the current subgroup.
+//
+// The shape to be loaded is split into the largest 2D loads supported
+// by the hardware.
```

Comment:
what happens to, say, 1d tensors?

Reply:
I don't know. I would assume it will crash in the exact same way as the current linalg-to-xegpu lowering does

<details><summary>An attempt to use linalg-to-xegpu pass with 1D tensors/memrefs on the current main branch</summary>

```
gc-opt: /home/jovyan/llvm/mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:83: static void mlir::xegpu::CreateNdDescOp::build(mlir::OpBuilder&, mlir::OperationState&, mlir::Type, mlir::TypedValue<mlir::MemRefType>, llvm::ArrayRef<mlir::OpFoldResult>): Assertion `ty.hasStaticShape() && offsets.size() == (size_t)ty.getRank()' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: ./bin/gc-opt /home/jovyan/graph-compiler/test/mlir/test/gc/Transforms/GPU/linalg-to-xegpu1d.mlir "-linalg-to-xegpu=dpas-tile=8,16,16 k-tile=16" -canonicalize -split-input-file
 #0 0x00005571ec59bb30 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (./bin/gc-opt+0x589bb30)
 #1 0x00005571ec598f3f llvm::sys::RunSignalHandlers() (./bin/gc-opt+0x5898f3f)
 #2 0x00005571ec599095 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007fd97d43f520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007fd97d4939fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007fd97d4939fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x00007fd97d4939fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x00007fd97d43f476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007fd97d4257f3 abort ./stdlib/abort.c:81:7
 #9 0x00007fd97d42571b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#10 0x00007fd97d436e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#11 0x00005571e931f149 mlir::xegpu::CreateNdDescOp::build(mlir::OpBuilder&, mlir::OperationState&, mlir::Type, mlir::detail::TypedValue<mlir::MemRefType>, llvm::ArrayRef<mlir::OpFoldResult>) (./bin/gc-opt+0x261f149)
#12 0x00005571e9a4de94 mlir::xegpu::CreateNdDescOp mlir::OpBuilder::create<mlir::xegpu::CreateNdDescOp, mlir::xegpu::TensorDescType&, mlir::detail::TypedValue<mlir::MemRefType>, llvm::SmallVector<mlir::OpFoldResult, 6u>&>(mlir::Location, mlir::xegpu::TensorDescType&, mlir::detail::TypedValue<mlir::MemRefType>&&, llvm::SmallVector<mlir::OpFoldResult, 6u>&) /home/jovyan/llvm/llvm-install-imex-17_oct/include/mlir/IR/Builders.h:517:22
#13 0x00005571e9a2af4e (anonymous namespace)::createDescriptorTiles(mlir::PatternRewriter&, mlir::Location, mlir::Value, llvm::ArrayRef<long>, llvm::ArrayRef<long>, llvm::ArrayRef<long>, int, bool) /home/jovyan/graph-compiler/lib/gc/Transforms/GPU/LinalgToXeGPU.cpp:661:0
#14 0x00005571e9a2b585 (anonymous namespace)::createCoarseDscTiles(mlir::PatternRewriter&, mlir::Location, mlir::Value, llvm::ArrayRef<long>, bool, bool) /home/jovyan/graph-compiler/lib/gc/Transforms/GPU/LinalgToXeGPU.cpp:735:0
#15 0x00005571e9a2fa31 (anonymous namespace)::createEltwiseKernel(mlir::linalg::LinalgOp, mlir::PatternRewriter&) /home/jovyan/graph-compiler/lib/gc/Transforms/GPU/LinalgToXeGPU.cpp:1321:0
#16 0x00005571e9a42f80 (anonymous namespace)::ConvertNamedEltwiseToXeGPU<mlir::linalg::AddOp>::matchAndRewrite(mlir::linalg::AddOp, mlir::PatternRewriter&) const /home/jovyan/graph-compiler/lib/gc/Transforms/GPU/LinalgToXeGPU.cpp:1494:0
#17 0x00005571e9a6590c mlir::detail::OpOrInterfaceRewritePatternBase<mlir::linalg::AddOp>::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const /home/jovyan/llvm/llvm-install-imex-17_oct/include/mlir/IR/PatternMatch.h:332:3
#18 0x00005571ec0c6bc8 mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<llvm::LogicalResult (mlir::Pattern const&)>) (./bin/gc-opt+0x53c6bc8)
#19 0x00005571ec08f3de (anonymous namespace)::GreedyPatternRewriteDriver::processWorklist() GreedyPatternRewriteDriver.cpp:0:0
#20 0x00005571ec091be5 mlir::applyPatternsAndFoldGreedily(mlir::Region&, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig, bool*) (./bin/gc-opt+0x5391be5)
#21 0x00005571e9915394 mlir::applyPatternsAndFoldGreedily(mlir::Operation*, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig, bool*) /home/jovyan/llvm/llvm-install-imex-17_oct/include/mlir/Transforms/GreedyPatternRewriteDriver.h:159:37
#22 0x00005571e9a30f6e (anonymous namespace)::LinalgToXeGPU::runOnOperation() /home/jovyan/graph-compiler/lib/gc/Transforms/GPU/LinalgToXeGPU.cpp:1649:0
#23 0x00005571ec1c3479 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (./bin/gc-opt+0x54c3479)
#24 0x00005571ec1c3931 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (./bin/gc-opt+0x54c3931)
#25 0x00005571ec1c3cd6 mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::'lambda'(mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo&)::operator()(mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo&) const Pass.cpp:0:0
#26 0x00005571ec1c29a5 mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool) (./bin/gc-opt+0x54c29a5)
#27 0x00005571ec1c3280 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (./bin/gc-opt+0x54c3280)
#28 0x00005571ec1c3931 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (./bin/gc-opt+0x54c3931)
#29 0x00005571ec1c4995 mlir::PassManager::run(mlir::Operation*) (./bin/gc-opt+0x54c4995)
#30 0x00005571e98c5217 performActions(llvm::raw_ostream&, std::shared_ptr<llvm::SourceMgr> const&, mlir::MLIRContext*, mlir::MlirOptMainConfig const&) MlirOptMain.cpp:0:0
#31 0x00005571e98c5c2c processBuffer(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::MlirOptMainConfig const&, mlir::DialectRegistry&, llvm::ThreadPoolInterface*) MlirOptMain.cpp:0:0
#32 0x00005571e98c5d8d llvm::LogicalResult llvm::function_ref<llvm::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>::callback_fn<mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&)::'lambda'(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>(long, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) MlirOptMain.cpp:0:0
#33 0x00005571ec467b1f mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<llvm::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, llvm::StringRef, llvm::StringRef)::'lambda'(llvm::StringRef)::operator()(llvm::StringRef) const ToolUtilities.cpp:0:0
#34 0x00005571ec468472 mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<llvm::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, llvm::StringRef, llvm::StringRef) (./bin/gc-opt+0x5768472)
#35 0x00005571e98bd56c mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&) (./bin/gc-opt+0x2bbd56c)
#36 0x00005571e98c5ef0 mlir::MlirOptMain(int, char**, llvm::StringRef, llvm::StringRef, mlir::DialectRegistry&) (./bin/gc-opt+0x2bc5ef0)
#37 0x00005571e98c6417 mlir::MlirOptMain(int, char**, llvm::StringRef, mlir::DialectRegistry&) (./bin/gc-opt+0x2bc6417)
#38 0x00005571e70d410c main /home/jovyan/graph-compiler/src/gc-opt/gc-opt.cpp:75:0
#39 0x00007fd97d426d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#40 0x00007fd97d426e40 call_init ./csu/../csu/libc-start.c:128:20
#41 0x00007fd97d426e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#42 0x00005571e70d3cf5 _start (./bin/gc-opt+0x3d3cf5)
Aborted (core dumped)
```

</details>

(1D is not supported by linalg-to-xegpu)